### PR TITLE
Csv upload error improvements

### PR DESCRIPTION
**Pull request by gibsonbailey**
@ckcollab 

# A brief description of the purpose of the changes contained in this PR.
I improved the csv upload process by implementing smarter header row recognition and a more-convenient error reporting method.

# Known issues to be addressed in a separate PR
...

## Note:
This change is running on production, because Isabelle was not able to upload the CSV of her students (~100 students) without it. She wasn't able to figure out what was wrong with the CSV, because the upload mechanism didn't tell her which row was breaking the upload. After uploading this change, she was able to upload her CSV. 

# A checklist for hand testing
- [x] Upload normal.csv (***WITHOUT*** lowercase `Last Name` header column).
- [x] Verify upload succeeded.
- [x] Delete students and teams from class.
- [x] Upload lowercase_last_name.csv (***WITH*** lowercase `last name` header column).
- [x] Verify upload succeeded.
- [x] Delete students and teams from class.
- [x] Upload invalid_email.csv (***WITH*** invalid email address in 2nd student row).
- [x] Verify upload did not succeed and a helpful error message was presented.

# Sample files in the zip below:
[sample_upload_csvs.zip](https://github.com/codalab/chagrade/files/4172633/sample_upload_csvs.zip)

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] Hand tested by reviewer
- [x] Ready to merge

### CODE REVIEW

File: apps/api/views/profiles.py [hunk 1]
```diff
@@ -44,14 +45,30 @@ class TestStudentViewSet(ModelViewSet):
     serializer_class = serializers.profiles.TestStudentSerializer
 
 
+def make_ordinal(n):
```
Comment (gibsonbailey) [inline apps/api/views/profiles.py:48]: Used to render the index of the student with an error to a human-readable ordinal string.

File: apps/api/views/profiles.py [hunk 2]
```diff
@@ -72,15 +89,19 @@ def create_students_from_csv(request, version):
...
-                    # If there's something in the student leader column
-                    if len(row) == 7 and contains_valid_team_name and row[6] in ['t', 'T', 'true', 'True', 'TRUE']:
-                        new_student.team.leader = new_student
-                        new_student.team.save()
+                try:
+                    new_student_serializer.is_valid(raise_exception=True)
+                except ValidationError:
+                    if new_student_serializer.errors:
+                        error_dict = dict(new_student_serializer.errors)
+                        error_dict['student'] = {'identifier': [ErrorDetail(f'From top of CSV, {make_ordinal(line_count + 1)} student with email: {str(row[4])}', code='invalid')]}
```
Comment (gibsonbailey) [inline apps/api/views/profiles.py]: Create descriptive error for the identification of the problem row in the CSV.

File: apps/api/views/profiles.py [hunk 3]
```diff
@@ -44,14 +45,30 @@ class TestStudentViewSet(ModelViewSet):
...
+
 @api_view(['POST'])
 def create_students_from_csv(request, version):
     with open(request.FILES['file'].temporary_file_path()) as csv_file:
         csv_reader = csv.reader(csv_file, delimiter=',')
         line_count = 0
         for row in csv_reader:
             if len(row) >= 6:
-                if row[0] == 'First Name' and row[3] == 'Student ID' and row[4] == 'Student Email':
+                if line_count == 0 and 'first name' in row[0].lower() and 'student id' in row[3].lower() and 'student email' in row[4].lower():
```
Comment (gibsonbailey) [inline apps/api/views/profiles.py:70]: Case insensitive header row check.

File: apps/homework/tasks.py [hunk 4]
```diff
@@ -134,5 +134,7 @@ def post_submission(submission_pk, data_file=None):
         )
         logger.info(f'Succeeded posting submission with id {submission.id} to phase.')
     else:
+        logger.info(phase_final_resp.status_code)
```
Comment (gibsonbailey) [inline apps/homework/tasks.py:138]: Extra little bug fix:
It's nice to see what Codalab responds with in the logs in the case of a failed submission POST.

File: apps/api/views/profiles.py [hunk 5]
```diff
@@ -44,14 +45,30 @@ class TestStudentViewSet(ModelViewSet):
...
+        make_ordinal(0)   => '0th'
+        make_ordinal(3)   => '3rd'
+        make_ordinal(122) => '122nd'
+        make_ordinal(213) => '213th'
+    '''
+    n = int(n)
+    suffix = ['th', 'st', 'nd', 'rd', 'th'][min(n % 10, 4)]
+    if 11 <= (n % 100) <= 13:
+        suffix = 'th'
+    return str(n) + suffix
```
Comment (ckcollab) [inline apps/api/views/profiles.py]: use f strings?

File: apps/api/views/profiles.py [hunk 6]
```diff
@@ -44,14 +45,30 @@ class TestStudentViewSet(ModelViewSet):
     serializer_class = serializers.profiles.TestStudentSerializer
 
 
+def make_ordinal(n):
+    '''
```
Comment (ckcollab) [inline apps/api/views/profiles.py]: docstring on same line as quotes, quotes should be double quotes

File: apps/api/views/profiles.py [hunk 7]
```diff
@@ -44,14 +45,30 @@ class TestStudentViewSet(ModelViewSet):
...
+    Convert an integer into its ordinal representation::
+
+        make_ordinal(0)   => '0th'
+        make_ordinal(3)   => '3rd'
+        make_ordinal(122) => '122nd'
+        make_ordinal(213) => '213th'
+    '''
+    n = int(n)
+    suffix = ['th', 'st', 'nd', 'rd', 'th'][min(n % 10, 4)]
+    if 11 <= (n % 100) <= 13:
```
Comment (ckcollab) [inline apps/api/views/profiles.py:58]: could write a few simple tests to cover this just-in-case

File: apps/api/views/profiles.py (hunk 2 above)
Comment (ckcollab) [reply apps/api/views/profiles.py]: why not just 
```
raise ValidationError(f'From top of CSV, {make_ordinal(line_count + 1)} student with email: {str(row[4])}')
```

File: apps/api/views/profiles.py [hunk 8]
```diff
@@ -72,15 +89,19 @@ def create_students_from_csv(request, version):
...
+                    new_student_serializer.is_valid(raise_exception=True)
+                except ValidationError:
+                    if new_student_serializer.errors:
+                        error_dict = dict(new_student_serializer.errors)
+                        error_dict['student'] = {'identifier': [ErrorDetail(f'From top of CSV, {make_ordinal(line_count + 1)} student with email: {str(row[4])}', code='invalid')]}
+                        raise ValidationError(error_dict)
+
+                new_student = new_student_serializer.save()
+                # If there's something in the student leader column
+                if len(row) == 7 and contains_valid_team_name and row[6] in ['t', 'T', 'true', 'True', 'TRUE']:
```
Comment (ckcollab) [inline apps/api/views/profiles.py]: convert to lower before comparison?

File: apps/homework/tasks.py (hunk 4 above)
Comment (ckcollab) [reply apps/homework/tasks.py:138]: cool

**Review by ckcollab — CHANGES_REQUESTED** [review]

File: apps/api/views/profiles.py (hunk 2 above)
Comment (gibsonbailey) [reply apps/api/views/profiles.py]: ✅

File: apps/api/views/profiles.py (hunk 6 above)
Comment (gibsonbailey) [reply apps/api/views/profiles.py]: ✅

File: apps/api/views/profiles.py (hunk 7 above)
Comment (gibsonbailey) [reply apps/api/views/profiles.py:58]: ✅

File: apps/api/views/profiles.py (hunk 8 above)
Comment (gibsonbailey) [reply apps/api/views/profiles.py]: ✅

File: apps/api/views/profiles.py (hunk 5 above)
Comment (gibsonbailey) [reply apps/api/views/profiles.py]: ✅

File: apps/api/tests/test_profiles_api.py [hunk 9]
```diff
@@ -56,9 +56,9 @@ def test_lowercase_header_column_csv_upload(self):
     def test_invalid_email_csv_upload(self):
         self.client.force_login(self.user)
         initial_student_count = self.klass.enrolled_students.count()
-        normal_csv_filename = os.path.join(os.path.dirname(settings.BASE_DIR), 'apps/api/tests/files/sample_upload_csvs/invalid_email_2nd_student.csv')
+        invalid_email_csv_filename = os.path.join(os.path.dirname(settings.BASE_DIR), 'apps/api/tests/files/sample_upload_csvs/invalid_email_2nd_student.csv')
 
-        with open(normal_csv_filename, 'r+') as f:
+        with open(invalid_email_csv_filename, 'r+') as f:
             resp = self.client.post(reverse('api:create_students_from_csv', kwargs={'version': 'v1'}), {'file': f, 'klass': self.klass.pk})
             assert resp.status_code == 400
```
Comment (ckcollab) [inline apps/api/tests/test_profiles_api.py:63]: should also check that the error message is set properly

**Review by ckcollab — APPROVED** [review]

**Pull request merged**